pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,290 @@
+	return &cachingUnstructuredExtractor{
+		dc:    dc,
+		cache: cache,
+	}, nil
+}
+
+// Extract extracts the applied configuration owned by fieldManager from an unstructured object.
+// Note that the apply configuration itself is also an unstructured object.
+func (e *cachingUnstructuredExtractor) Extract(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "")
+}
+
+// ExtractStatus is the same as ExtractUnstructured except
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+// schemerForGVK queries the openapi v3 discovery endpoint
+// and builds a custom openapi.GroupVersion to retrieve the schema
+// only for the GroupVersion of the given GroupVersionKind
+// It also stores the returned ETag associated with the GV for caching
+func (e *cachingUnstructuredExtractor) schemerForGVK(gvk schema.GroupVersionKind) (*CustomOAPIGroupVersion, error) {
+	data, err := e.dc.RESTClient().Get().
+		AbsPath("/openapi/v3").
+		Do(context.TODO()).
+		Raw()
+
+	if err != nil {
+		return nil, err
+	}
+
+	discoMap := &handler3.OpenAPIV3Discovery{}
+	err = json.Unmarshal(data, discoMap)
+	if err != nil {
+		return nil, err
+	}
+
+	gvPath := gvkToRelativeAPIPath(gvk)
+	oapiGV, ok := discoMap.Paths[gvPath]
+	if !ok {
+		return nil, fmt.Errorf("cannot find GroupVersion %q in discovery", gvPath)
+	}
+
+	parse, err := url.Parse(oapiGV.ServerRelativeURL)
+	if err != nil {
+		return nil, err
+	}
+	useClientPrefix := strings.HasPrefix(oapiGV.ServerRelativeURL, "/openapi/v3")
+	etag := parse.Query().Get("hash")
+
+	return NewCustomOAPIGroupVersion(e.dc.RESTClient(), oapiGV, useClientPrefix, etag), nil
+}
+
+// gvkToRelativeAPIPath constructs the OpenAPI path for the given GVK
+func gvkToRelativeAPIPath(gvk schema.GroupVersionKind) string {
+	if gvk.Group == "" {
+		return "api/" + gvk.Version
+	}
+	return "apis/" + gvk.GroupVersion().String()
+}
+
+// parserFnForSchemer returns a BuildParserFn that builds a new *GvkParser
+// from the OpenAPI GroupVersion schema client
+func parserFnForSchemer(schemer *CustomOAPIGroupVersion) func() (*GvkParser, error) {
+	return func() (*GvkParser, error) {
+		s, err := schemer.Schema("application/json")
+		if err != nil {
+			return nil, err
+		}
+		var oapi spec3.OpenAPI
+		if err := json.Unmarshal(s, &oapi); err != nil {
+			return nil, err
+		}
+
+		specs := map[string]*spec.Schema{}
+		for k, v := range oapi.Components.Schemas {
+			specs[k] = v
+		}
+		return NewGVKParser(specs, false)
+	}
+}
+
+func (e *cachingUnstructuredExtractor) objectForGVK(gvk schema.GroupVersionKind) (*typed.ParseableType, error) {
```

Comment:
if this function stays quite simple like this, you can maybe even inline it into `extractUnstructured` so it's easier to read the logic, unless otherwise is needed for e.g. unit tests.